doc/config.rst

=============
Configuration
=============

Billometer configuration has tree main parts:

* Base Django Configuration (Database, Auth, ..)
* Keystone
* Billometer Resources
* Billometer Worker (Worker tasks, Custom Task)

Base
----

.. code-block:: python

    ALLOWED_HOSTS = ['10.10.10.70']

    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.postgresql_psycopg2',
            'PORT': '5432',
            'HOST': '127.0.0.1',
            'NAME': 'billometer',
            'PASSWORD': 'password',
            'USER': 'billometer'
        }
    }

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
            'LOCATION': '127.0.0.1:11211',
            'TIMEOUT': 120,
            'KEY_PREFIX': 'CACHE_BILLOMETER'
        }
    }

Keystone
--------

.. code-block:: python

    BROKER_URL = 'amqp://billometer:sfslkfsfe@hostname.cz:5672//billometer'

    KEYSTONE_REGION = "regionOne"

    KEYSTONE_SERVICE_TOKEN = "service token"

    KEYSTONE_USER = "billometer"

    KEYSTONE_PASSWORD = "password"

    KEYSTONE_SERVICE_ENDPOINT = "http://10.10.10.70:35357/v2.0"

    OPENSTACK_KEYSTONE_URL = KEYSTONE_SERVICE_ENDPOINT

    OPENSTACK_SSL_NO_VERIFY = True

    OPENSTACK_API_VERSIONS = {
        'identity': 2.0
    }


Billometer Resources
--------------------

.. code-block:: python

    RESOURCE_PRICE = {
        'cinder.volume': {
            '7k2_SAS': '0.008205',
            '10k_SAS': '0.027383',
            '15k_SAS': '0.034232',
            'EasyTier': '0.041082',
        },
        'nova.memory': '0.000369',
        'nova.cpu': '0.821904',
        'neutron.floating_ip': '0.136972',
        'glance.image': '0.002739',
    }

    BILLING_EXTRA_RESOURCES = {
        'network.rx': {
            'threshold': '150000',
            'price_rate': '0.000002222',
            'resource': 'network.rx',
            'label': 'Network RX'
        },
        'network.tx': {
            'threshold': '99999999999999',
            'price_rate': '0.000002222',
            'resource': 'network.tx',
            'label': 'Network TX'
        },
        '7k2_SAS': {
            'price_rate': '0.008205',
            'resource': 'cinder.volume',
            'name': '7k2_SAS',
            'label': '7k2 SAS'
        },
        '10k_SAS': {
            'price_rate': '0.027383',
            'resource': 'cinder.volume',
            'label': '10k2 SAS',
            'name': '10k_SAS',
        },
        '15k_SAS': {
            'price_rate': '0.034232',
            'resource': 'cinder.volume',
            'label': '15k2 SAS',
            'name': '15k_SAS',
        },
        'EasyTier': {
            'price_rate': '0.041082',
            'resource': 'cinder.volume',
            'label': 'Easy Tier',
            'name': 'EasyTier',
        },

    }

Billometer Worker
-----------------

In default state Billometer is configured to ``sync_all`` and ``collect_all`` which means base resource such as:

* nova - instances
* glance - images
* cinder - volumes
* neutron - ip addresses

.. code-block:: python

    CELERYBEAT_SCHEDULE = {
        'collect_all': {
            'task': 'billometer.tasks.collect_all',
            'schedule': timedelta(seconds=3600),
            'args': tuple()
        },
        'sync_all': {
            'task': 'billometer.tasks.sync_all',
            'schedule': timedelta(seconds=3600),
            'args': tuple()
        },
    }

For network IN/OUT billing enable these tasks

.. code-block:: python

    CELERYBEAT_SCHEDULE = {

        ...

        'collect_network': {
            'task': 'billometer.tasks.network.collect_network',
            'schedule': timedelta(seconds=3600),
            'args': tuple()
        },
        'sync_network': {
            'task': 'billometer.tasks.network.sync_network',
            'schedule': timedelta(seconds=3600),
            'args': tuple()
        }
    }

.. note::

    Billometer works with 1 hours period.


debian/changelog
billometer (0.1) unstable; urgency=low

  * initial release

 -- Michael Kuty <michael.kuty@tcpcloud.eu>  Tue, 01 Sep 2015 20:00:42 +0200


billometer/settings.py
# -*- coding: utf-8 -*-

from os.path import join, dirname, abspath, normpath

ALLOWED_HOSTS = ['*']

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': '/var/lib/billometer/db.sqlite3',
    }
}

OPENSTACK_SSL_NO_VERIFY = True

OPENSTACK_API_VERSIONS = {
    'identity': 2.0
}

USE_TZ = True

OPENSTACK_KEYSTONE_MULTIDOMAIN_SUPPORT = False
OPENSTACK_KEYSTONE_DEFAULT_DOMAIN = 'Default'

DEBUG = False
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    ('Admin', 'mail@newt.cz'),
)

MANAGERS = ADMINS

SITE_ID = 1
SITE_NAME = 'billometer'

TIME_ZONE = 'Europe/Prague'

LANGUAGE_CODE = 'en'

LANGUAGES = (
    ('en', 'EN'),
)

USE_I18N = True

MEDIA_URL = '/media/'
STATIC_URL = '/static/'

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.contrib.auth.context_processors.auth',
    'django.core.context_processors.debug',
    'django.core.context_processors.i18n',
    'django.core.context_processors.media',
    'django.core.context_processors.request',
    'django.core.context_processors.static',
)

TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

ROOT_URLCONF = 'billometer.urls'

INSTALLED_APPS = (
    'django',
    'django_extensions',
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.admin',
    'django.contrib.admindocs',
    'django.contrib.staticfiles',
    'rest_framework',
    'openstack_auth',
    'billometer',
    'djcelery',
)

RESOURCE_PRICE = {
    'cinder.volume': {
        '7k2_SAS': '0.008205',
        '10k_SAS': '0.027383',
        '15k_SAS': '0.034232',
        'EasyTier': '0.041082',
    },
    'nova.memory': '0.000369',
    'nova.cpu': '0.821904',
    'neutron.floating_ip': '0.136972',
    'glance.image': '0.002739',
}

BILLING_EXTRA_RESOURCES = {
    'network.rx': {
        'threshold': '150000',
        'price_rate': '0.000002222',
        'resource': 'network.rx',
        'label': 'Network RX'
    },
    'network.tx': {
        'threshold': '99999999999999',
        'price_rate': '0.000002222',
        'resource': 'network.tx',
        'label': 'Network TX'
    },
}

STATICFILES_FINDERS = (
    "django.contrib.staticfiles.finders.FileSystemFinder",
    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
)

LOGIN_URL = '/admin/login/'
LOGIN_REDIRECT_URL = '/admin/'

AUTHENTICATION_BACKENDS = (
    'django.contrib.auth.backends.ModelBackend',
    'openstack_auth.backend.KeystoneBackend',
)

REST_FRAMEWORK = {
    'DEFAULT_MODEL_SERIALIZER_CLASS':
        'rest_framework.serializers.HyperlinkedModelSerializer',
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.AllowAny'
    ]
}

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'root': {
        'level': 'WARNING',
        'handlers': ['console'],
    },
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'formatters': {
        'verbose': {
            'format': "[%(asctime)s] %(levelname)s [%(name)s:%(lineno)s] %(message)s",
            'datefmt': "%d/%b/%Y %H:%M:%S"
        },
        'simple': {
            'format': '%(levelname)s %(message)s'
        },
    },
    'handlers': {
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'verbose'
        },
    },
    'loggers': {
        'django.request': {
            'handlers': ['console'],
            'level': 'DEBUG',
            'propagate': True,
        },
    }
}

try:
    with open("/etc/billometer/settings.py") as f:
        code = compile(f.read(), "/etc/billometer/settings.py", 'exec')
        exec(code)
except IOError:
    pass

if hasattr(globals(), "LOCAL_INSTALLED_APPS"):
    INSTALLED_APPS += LOCAL_INSTALLED_APPS
